sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/thetasketchesExpressions.scala

```diff
@@ -0,0 +1,321 @@
+
+  override def prettyName: String = "theta_sketch_estimate"
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(BinaryType)
+
+  override def dataType: DataType = LongType
+
+  override def nullSafeEval(input: Any): Any = {
+    val buffer = input.asInstanceOf[Array[Byte]]
+
+    val memory = try {
+      Memory.wrap(buffer)
+    } catch {
+      case _: IllegalArgumentException | _: IndexOutOfBoundsException =>
+        throw QueryExecutionErrors.thetaInvalidInputSketchBuffer(prettyName)
+    }
+
+    val sketch = try {
+      CompactSketch.wrap(memory)
+    } catch {
+      case _: SketchesArgumentException =>
+        throw QueryExecutionErrors.thetaInvalidInputSketchBuffer(prettyName)
+    }
+
+    Math.round(sketch.getEstimate)
+  }
+}
+
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = """
+    _FUNC_(first, second, lgNomEntries) - Merges two binary representations of
+    Datasketches ThetaSketch objects, using a Datasketches Union object. Set
+    lgNomEntries to a value between 4 and 26 to find the unions of sketches with different
+    union buffer sizes values (defaults to 12). """,
+  examples = """
+    Examples:
+      > SELECT theta_sketch_estimate(_FUNC_(theta_sketch_agg(col1), theta_sketch_agg(col2))) FROM VALUES (1, 4), (1, 4), (2, 5), (2, 5), (3, 6) tab(col1, col2);
+       6
+  """,
+  group = "misc_funcs",
+  since = "4.1.0")
+// scalastyle:on line.size.limit
+case class ThetaUnion(first: Expression, second: Expression, third: Expression)
+    extends TernaryExpression
+    with CodegenFallback
+    with ExpectsInputTypes {
+  override def nullIntolerant: Boolean = true
+
+  def this(first: Expression, second: Expression) = {
+    this(first, second, Literal(ThetaSketchUtils.DEFAULT_LG_NOM_LONGS))
+  }
+
+  def this(first: Expression, second: Expression, third: Int) = {
+    this(first, second, Literal(third))
+  }
+
+  override protected def withNewChildrenInternal(
+      newFirst: Expression,
+      newSecond: Expression,
+      newThird: Expression): ThetaUnion =
+    copy(first = newFirst, second = newSecond, third = newThird)
+
+  override def prettyName: String = "theta_union"
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(BinaryType, BinaryType, IntegerType)
+
+  override def dataType: DataType = BinaryType
+
+  override def nullSafeEval(value1: Any, value2: Any, value3: Any): Any = {
+    val logNominalEntries = value3.asInstanceOf[Int]
+    ThetaSketchUtils.checkLgNomLongs(logNominalEntries)
+
+    val memory1 = try {
+      Memory.wrap(value1.asInstanceOf[Array[Byte]])
+    } catch {
+      case _: IllegalArgumentException | _: IndexOutOfBoundsException =>
+        throw QueryExecutionErrors.thetaInvalidInputSketchBuffer(prettyName)
+    }
+
+    val sketch1 = try {
+      CompactSketch.wrap(memory1)
+    } catch {
+      case _: SketchesArgumentException =>
+        throw QueryExecutionErrors.thetaInvalidInputSketchBuffer(prettyName)
+    }
+
+    val memory2 = try {
```

Comment:
This code from here until L148 seems like a dup of L122-L134 above. Should we dedup it into a shared helper?

Same for L218 and L300 below. Could we make just one helper that we reuse 2 times for each relevant function in this file?

Reply:
Yep. What do you think of a function `def wrapSketch(buffer: Array[Byte], prettyName: String): CompactSketch` in the ThetaSketchUtils class?

Reply:
Sounds good to me!